### PR TITLE
Added :class:`.ChangeSpeed`, an animation wrapper that allows to smoothly change the speed at which an animation is played

### PR DESCRIPTION
**Pull request by sparshg**
## Overview

Closed #2031 

Speed of given animation can be changed by wrapping the animation with `ChangeSpeed` and passing {percentage of `run_time` : speed factor} to `speedinfo`. To increase/decrease the speed, the `run_time` of the animation will be changed accordingly. 

To apply speed change to updaters, you can use `ChangeSpeed.add_updater(mob, updater_func)` and wrap `Wait()` with `ChangeSpeed` to get changed `dt`. See examples below
 
## Explanation

- Fixed `rate_func` not working when passed to `AnimationGroup` or `Succession`. `rate_func` will spread over all the animations as expected.
- Added `ChangeSpeed` class which generates a `rate_func` according to the the `speedinfo` and a `speed_modifier` function and assigns to the passed animation
- `speed_modifier` function follows these condition: **f(0)=0, f'(0)=m, f'(f<sup>-1</sup>(1)) = n** where m is the initial speed, n is the final speed. A vertical parabola following these conditions is the default function: $$\left(n^{2}-m^{2} \over {4}\right)x^{2}+mx$$
- `f_inv_1` function is f<sup>-1</sup>(1) for `speed_modifier` which comes out to be this for above function: $$2 \over m+n$$
- ~Added `ChangedWait`  animation, modified version of `Wait` which follows `rate_func`~

### Example animations

See on youtube:

[](http://youtube.com/watch?v=6Ggcfgt30aw)

https://user-images.githubusercontent.com/43041139/133597529-8563a00e-8900-4cbc-9d58-064270172230.mp4

### Example 1

https://user-images.githubusercontent.com/43041139/171675375-46209df9-695c-4435-9ec0-8fa8483a62af.mp4

```py
def construct(self):
    a = Dot().shift(LEFT * 4)
    b = Dot().shift(RIGHT * 4)
    self.add(a, b)
    self.play(
        ChangeSpeed(
            AnimationGroup(
                a.animate(run_time=1).shift(RIGHT * 8),
                b.animate(run_time=1).shift(LEFT * 8),
            ),
            speedinfo={0.3: 1, 0.4: 0.1, 0.6: 0.1, 1: 1},
            rate_func=linear,
        )
    )
```

---

### Example 2

https://user-images.githubusercontent.com/43041139/171675381-b0992753-31a5-4443-8d35-1edc1ac9c784.mp4

```py
def construct(self):
    a = Dot().shift(LEFT * 4)
    self.add(a)

    ChangeSpeed.add_updater(a, lambda x, dt: x.shift(RIGHT * 4 * dt))
    self.play(
        ChangeSpeed(
            Wait(2),
            speedinfo={0.4: 1, 0.5: 0.2, 0.8: 0.2, 1: 1},
            affects_speed_updaters=True,
        )
    )
```

---

### Example 3

https://user-images.githubusercontent.com/43041139/171675383-48bdcc8e-eefb-4a2b-9d76-0a5e8d5bede4.mp4

```py
def construct(self):
    a = Dot().shift(LEFT * 4)
    self.add(a)

    ChangeSpeed.add_updater(a, lambda x, dt: x.shift(RIGHT * 4 * dt))
    self.wait()
    self.play(
        ChangeSpeed(
            Wait(),
            speedinfo={1: 0},
            affects_speed_updaters=True,
        )
    )
```

Animation speed can be finely controlled to make effects such as slow-mo.

## Reviewer Checklist
- [ ] The PR title is descriptive enough for the changelog, and the PR is labeled correctly
- [ ] If applicable: newly added non-private functions and classes have a docstring including a short summary and a PARAMETERS section
- [ ] If applicable: newly added functions and classes are tested

### CODE REVIEW

**Review by behackl — COMMENTED** [review]
Thank you for your reworked contribution! This is indeed much cleaner than the last approach, I like it quite a lot!

I have not fully read the code yet (but am mainly just missing the static `add_updater` method, which seems a bit odd to me, but I think I understand the idea behind it), but from what I have read I think this is in a pretty good state already. Take a look at my comments for more detailed remarks. :-)

File: manim/animation/speedmodifier.py [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+            prevnode = node
+            m = n
+
+        def func(x):
+            newx = piecewise(
+                self.rate_func(x),
+                [condition(self.rate_func(x)) for condition in self.conditions],
+                self.functions,
+            )
+            ChangeSpeed.dt = (newx - self.t) * self.anim.run_time
```
[the remark this answers is not in the document]
Comment (sparshg) [reply manim/animation/speedmodifier.py]: I didn't consider several ChangeSpeed animations playing at once. Even if many animations play, there can only be single `dt` for the scene/updaters. So, we have to explicitly define which animation is going to control `dt` in such a case.  Only one `ChangeSpeed` animation at a time can have this ability which is why I added a class variable `is_changing_dt`. I added a bool parameter `affects_speed_updaters` which if true, can control `ChangeSpeed.dt`. If multiple animations try to do so together, I can detect that using `is_changin_dt` class variable leading to assertion error

**Comment by sparshg** [comment]
Thanks for the review, I guess I addressed all of your comments and fixed the bugs I could find.

File: manim/animation/speedmodifier.py [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+        self.anim.begin()
+
+    def clean_up_from_scene(self, scene: Scene) -> None:
+        self.anim.clean_up_from_scene(scene)
+
+    def _setup_scene(self, scene) -> None:
+        self.anim._setup_scene(scene)
+
+
+class ChangedWait(Wait):
```
[the remark this answers is not in the document]
Comment (sparshg) [reply manim/animation/speedmodifier.py]: Thanks, this one was really useful.

**Review by kilacoda — APPROVED** [review]

**Review by behackl — CHANGES_REQUESTED** [review]
Sorry for the long time since my last review -- I'm very happy with this PR at this point! The last few issues I see with this:

- I've left a bunch of suggestions concerning formatting of the docstrings, please take a look,
- `ChangeSpeed.add_updater` *really* should have a docstring to explain what the method is used for. I don't think that rendered examples are necessary, but it should definitely have a short description,
- and finally, new functionality should be tested: I'd recommend copy-pasting the code for `SpeedModifierExample` to a test (I can help with that in case you don't want to look into how tests work; just let me know in case.

Other than that this seems good to go; pretty cool stuff -- thanks for your efforts!

**Review by behackl — APPROVED** [review]
I've left one final suggestion to further improve the docstring of `ChangeSpeed.add_updater`, but other than that I have no further comments. Looks good to me, very cool!

**Pull request merged**